src/Microsoft.Health.Dicom.Core/Features/Retrieve/IInstanceStore.cs

```diff
@@ -30,10 +31,14 @@ public interface IInstanceStore
         /// <summary>
         /// Gets idenfiers of instances within the given range of watermarks.
         /// </summary>
-        /// <param name="watermarkRange"></param>
+        /// <param name="watermarkRange">The watermark range</param>
+        /// <param name="indexStatus">The index status</param>
         /// <param name="cancellationToken">The cancellation token</param>
         /// <returns>The instanceidentifiers</returns>
-        Task<IEnumerable<VersionedInstanceIdentifier>> GetInstanceIdentifiersAsync(WatermarkRange watermarkRange, CancellationToken cancellationToken = default);
+        Task<IReadOnlyList<VersionedInstanceIdentifier>> GetInstanceIdentifiersByWatermarkRangeAsync(
+            WatermarkRange watermarkRange,
+            IndexStatus indexStatus,
```

Comment:
In what scenario would we want to index data that is still pending creation? Presumably we have a read lock on the entire `[dbo].[ExtendedQueryTag]` table when inserting an instance so that we correctly see all of the DICOM instances

Reply:
I cannot image scenarios except for testing (check test `GivenInstances_WhenGetInstanceIdentifiersByWatermarkRange_ThenItShouldReturnInstancesInRange`) -- in integration test, we call `InstanceStore.CreateInstanceIndexAsync` to create index on test instance , which set indexStatus as `Creating`. 

Due to the upgrade difficulty of SQL schema, I think allowing it supporting filtering by IndexStatus is a little effort but add much flexibility.